reputation/src/trs.rs

```diff
@@ -0,0 +1,747 @@
+        assert_eq!(a.get(&id1), Reputation(95));
+        assert_eq!(a.penalize(&id1, cnst(10)), Ok(Reputation(10)));
+        assert_eq!(a.expire(&Alpha(11)), Reputation(30));
+        assert_eq!(a.get(&id1), Reputation(55));
+        assert_eq!(a.penalize(&id1, cnst(10)), Ok(Reputation(10)));
+        assert_eq!(a.expire(&Alpha(12)), Reputation(15));
+        assert_eq!(a.get(&id1), Reputation(30));
+    }
+
+    #[test]
+    fn penalize_overflow() {
+        // An error is returned when the penalize function returns an invalid
+        // new reputation
+        let mut a = Trs::new();
+        let id1 = "Alice".to_string();
+        assert_eq!(a.get(&id1), Reputation::default());
+        a.gain(Alpha(10), vec![(id1.clone(), Reputation(50))]);
+        assert_eq!(a.get(&id1), Reputation(50));
+        // Penalize this identity with more reputation that it currently has
+        assert!(a.penalize(&id1, |_| Reputation(1000)).is_err());
+    }
+
+    #[test]
+    fn multiple_identities() {
+        let mut a = Trs::new();
+        let id1 = "Alice".to_string();
+        let id2 = "Bob".to_string();
+        let id3 = "Eve".to_string();
+        // Helper function to be used with assert_eq
+        let reps =
+            |a: &Trs<String, Reputation, Alpha, _>| (a.get(&id1).0, a.get(&id2).0, a.get(&id3).0);
+        let v4 = vec![
+            (id1.clone(), Reputation(1024)),
+            (id2.clone(), Reputation(1024)),
+            (id3.clone(), Reputation(1024)),
+            (id2.clone(), Reputation(1024)),
+        ];
+        let v5 = vec![
+            (id1.clone(), Reputation(1024)),
+            (id3.clone(), Reputation(1024)),
+        ];
+        let v6 = vec![
+            (id1.clone(), Reputation(1024)),
+            (id2.clone(), Reputation(1024)),
+        ];
+        let v7 = vec![
+            (id1.clone(), Reputation(1024)),
+            (id2.clone(), Reputation(1024)),
+            (id3.clone(), Reputation(1024)),
+        ];
+        let v8 = vec![
+            (id1.clone(), Reputation(1024)),
+            (id2.clone(), Reputation(1024)),
+            (id3.clone(), Reputation(1024)),
+        ];
+        let v9 = vec![
+            (id1.clone(), Reputation(1024)),
+            (id2.clone(), Reputation(1024)),
+            (id3.clone(), Reputation(1024)),
+        ];
+        let v10 = vec![];
+        let v11 = vec![];
+        let v12 = vec![];
+        let v13 = vec![
+            (id1.clone(), Reputation(1024)),
+            (id2.clone(), Reputation(1024)),
+            (id3.clone(), Reputation(1024)),
+        ];
+        assert_eq!(reps(&a), (0, 0, 0));
+        a.expire(&Alpha(0));
+        a.gain(Alpha(4), v4);
+        assert_eq!(reps(&a), (1024, 2048, 1024));
+        let p = a.penalize_many(vec![(&id1, dmrg(1)), (&id2, dmrg(1)), (&id3, dmrg(1))]);
+        assert_eq!(p, Ok(Reputation(512 + 1024 + 512)));
+        assert_eq!(reps(&a), (512, 1024, 512));
+        a.expire(&Alpha(1));
+        a.gain(Alpha(5), v5);
```

Comment:
Could be a good option to add an assert_eq after gain calls to see how increments repurtation before to penalize it
assert_eq!(reps(&a), (1536 1024, 1536));